### PR TITLE
Bug 60274

### PR DESCRIPTION
**Pull request by blindio**

### CODE REVIEW

**Comment by blindio** [comment]
All build and test errors that I believe that I introduced  are resolved. Please tell me if this is incorrect.

**Comment by pmouawad** [comment]
Thanks for this contribution !
We will review it ASAP and give you feedback.
Regards
Philippe M.

**Comment by blindio** [comment]
I look forward to your feedback

**Comment by blindio** [comment]
Hi Philippe-

I have not heard anything for a while and wanted to know if you had an idea when someone would be able to review this.

Thanks,

Thorson

**Comment by pmouawad** [comment]
Hi Thorson,
Sorry for late reply, we are in the middle of 3.1 release preparation and I was on holiday.
Thanks again for you PR.
I started reviewing, few notes:

- WeightedDistributionController#initValueReplacer() => GuiPackage.getInstance() is null in Non GUI mode which is the expected mode for running load test. So if this is really needed (TestElement should avoid using GuiPackage.getInstance()) , the test should be done on nullity of GuiPackage.getInstance() to avoid Exception throwing and cached

- Same note for WeightedDistributionController#getNode()


RandomIntegerGenerator:

- When seed is not set, using ThreadLocalRandom.current() might be better

SubControllerIterator:

- Vector must be avoided

determineCurrentTestElement:

- Why do you test currSubController.isEnabled() ? Could you improve javadoc to make the algo clear for me ?

-Regarding code style, I'd prefer to have each class in its own file or static internal rather than having multiple classes in the same file (WeightedDistributionController => RandomIntegerGenerator, IntegerGenerator), SubControllerIterator might be better internal static ?

**Comment by blindio** [comment]
Thank you Philippe, I will make these changes in the next day or so

**Comment by pmouawad** [comment]
Hi Again,
For info, in your commit, no need to include the HTML files, only XML for documentation.
Thanks

**Comment by pmouawad** [comment]
Hi @blindio ,
Did you have time working on the reported issues ?
Thanks

**Comment by blindio** [comment]
Not yet, things got hectic for me at work, will resolve in a little while

File: src/components/org/apache/jmeter/control/WeightedDistributionController.java [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+            }
+        } else {
+            nextSubController = null;
+        }
+    }
+}
+
+/**
+ * This abstraction is necessary for unit tests, in which case a deterministic
+ * generator replaces the random one
```
Comment (vlsi) [inline src/components/org/apache/jmeter/control/WeightedDistributionController.java]: Why can't seed be used in order to make test consistent?
There are at least two ways:
1) Hard-code seed value right into the test
2) Use random seed, and enable configuring the seed from the command line. That way if a test with some specific seed value fails, we can reproduce it by supplying that particular value.

File: src/components/org/apache/jmeter/control/WeightedDistributionController.java [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+    }
+
+    /**
+     * Gets the cumulative probability.
+     *
+     * @return the cumulative probability
+     */
+    public int getCumulativeProbability() {
+        // recalculate if reset flag is set
+        if (resetCumulativeProbability) {
```
Comment (vlsi) [inline src/components/org/apache/jmeter/control/WeightedDistributionController.java]: Can you clarify when weight is recalculated?
The documentation suggests that weight is recalculated at run-time, however the flag `resetCumulativeProbability` is reset at design time only. That is it looks like weight distribution would not account for the variable changes in the runtime. Is that the case? Can you make documentation/intention more clear?

**Comment by vlsi** [comment]
While being able to enable-disable and rename the children from WDC UI is nice, I wonder how many worms are hidden in that "keep WDC and children tree" consistent can.


Nested "Weighted Distribution Controller" elements does not seem to work.

**Comment by blindio** [comment]
The reason that I do not rely on the seed is because, as I understand it,
you can count on seed to be consistent within am implementation of random,
but not across implementations.

Users of 3rd party JVMs may have consistent failures in the unit tests
because their implementation of Random provides a different distribution
than those using the Open/Oracle JVM.

Since the intention of the test is to determine if all potential random
numbers bin correctly, the systematic approach guarantees correct binning.

Furthermore, magic numbers in unit tests should be avoided if at all
possible.

-Thorson

**Comment by blindio** [comment]
That is strange about the nesting not working, given that I manually tested
nesting, and it is one of the unit tests as well. I will investigate.

**Comment by vlsi** [comment]
Just in case, C and C1 has the following groovy post-processor: `vars.put("x", String.valueOf(1+Integer.parseInt(vars.get("x"))))`

**Comment by blindio** [comment]
Regarding the run time variable changes, that is a really good point.  I
will test and consider a solution.

Thank you,

Thorson

**Comment by vlsi** [comment]
> I will test and consider a solution

The question is if that is feasible at all to support run-time variable changes and keep the performance sane

**Comment by blindio** [comment]
Agreed, I need to find some time to consider this issue.  I may be that the
variable resolution is not feasible without major changes.

Again, thank you,

Thorson

**Comment by pmouawad** [comment]
It appears similar feature is now available a 3rd party plugin :

* https://jmeter-plugins.org/?search=Weighted

So I propose to close this one.

**Pull request closed without merging**